### PR TITLE
RavenDB-18480 - Sharding - Revert Revisions View Endpoints

### PR DESCRIPTION
**Pull request by lastav5**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18480

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Processors/Revisions/AbstractRevisionsHandlerProcessorForRevertRevisions.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            var token = RequestHandler.CreateTimeLimitedOperationToken();
+            var operationId = RequestHandler.GetLongQueryString("operationId", required: false) ?? GetNextOperationId();
+
+            using (ServerStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+            await using (var writer = new AsyncBlittableJsonTextWriter(context, RequestHandler.ResponseBodyStream()))
+            {
+                writer.WriteOperationIdAndNodeTag(context, operationId, ServerStore.NodeTag);
+            }
+
+            RevertRevisions(operationId, configuration, token);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Revisions/AbstractRevisionsHandlerProcessorForRevertRevisions.cs]: call it `ScheduleRevertRevisions`

File: src/Raven.Server/Documents/Handlers/Processors/Revisions/AbstractRevisionsHandlerProcessorForRevertRevisions.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            {
+                var json = await context.ReadForMemoryAsync(RequestHandler.RequestBodyStream(), "revisions/revert");
+
+                configuration = JsonDeserializationServer.RevertRevisions(json);
+            }
+
+            var token = RequestHandler.CreateTimeLimitedOperationToken();
+            var operationId = RequestHandler.GetLongQueryString("operationId", required: false) ?? GetNextOperationId();
+
+            using (ServerStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Revisions/AbstractRevisionsHandlerProcessorForRevertRevisions.cs:38]: first schedule the operation, then write to the response

File: src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{
+    internal class RevisionsHandlerProcessorForRevertRevisions : AbstractRevisionsHandlerProcessorForRevertRevisions<DatabaseRequestHandler, DocumentsOperationContext>
+    {
+        public RevisionsHandlerProcessorForRevertRevisions([NotNull] DatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override void RevertRevisions(long operationId, RevertRevisionsRequest configuration, OperationCancelToken token)
+        {
+            var t = RequestHandler.Database.Operations.AddLocalOperation(
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs:18]: don't you need to do the continuation and dispose the token? Check here: https://github.com/ravendb/ravendb/blob/v6.0/src/Raven.Server/Documents/Sharding/Handlers/Processors/Queries/AbstractShardedOperationQueriesHandlerProcessor.cs#L59

File: src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+                OperationType.DatabaseRevert,
+                $"Revert database '{RequestHandler.Database.Name}' to {configuration.Time} UTC.",
+                detailedDescription: null,
+                onProgress => RequestHandler.Database.DocumentsStorage.RevisionsStorage.RevertRevisions(configuration.Time, TimeSpan.FromSeconds(configuration.WindowInSec), onProgress, token),
+                token: token);
+        }
+
+        protected override long GetNextOperationId()
+        {
+            return ServerStore.Operations.GetNextOperationId();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs]: you want to get the operation Id from the RequestHandler.Database.Operations.GetNextOperationId(), not from the server one.

Comment (ppekrol) [reply src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs]: I see it was like that in the old code, this is a mistake. We should fix it in older versions as well.

File: src/Raven.Server/Documents/Revisions/RevertRevisionsOperation.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+using System.Net.Http;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Operations;
+using Raven.Client.Http;
+using Raven.Client.Json;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Revisions
+{
+    public class RevertRevisionsOperation : IMaintenanceOperation<OperationIdResult>
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Revisions/RevertRevisionsOperation.cs]: IIRC there was a commands namespace move it there

File: src/Raven.Server/Documents/Handlers/Processors/Revisions/AbstractRevisionsHandlerProcessorForRevertRevisions.cs (hunk 1 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Revisions/AbstractRevisionsHandlerProcessorForRevertRevisions.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs (hunk 3 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs:18]: this seems to already happen inside? https://github.com/ravendb/ravendb/blob/ffe03341cb20df1a9ac8e1d4b187bd61909de88c/src/Raven.Server/Documents/Operations/AbstractOperations.cs#L71

File: src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForRevertRevisions.cs]: done

File: src/Raven.Server/Documents/Revisions/RevertRevisionsOperation.cs (hunk 5 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Revisions/RevertRevisionsOperation.cs]: moved to `Raven.Server.Documents.Operations`

**Review by ppekrol — APPROVED** [review]

**Pull request merged**